Fix unsafe type usages - Issue #663 & use of params - Issue #610

ArcEye opened the pull request:
in instantiated components.

instcomp continues to use params but now prints a warning

Signed-off-by: Mick <arceye@mgware.co.uk>

strahlex commented on src/hal/i_components/biquad.icomp:
```diff
@@ -122,8 +122,8 @@ Biquad_CalcCoeffs(void *arg, unsigned long period)
 
-    double                      sampleRate, w0, alpha;
-    double                      b0, b1, b2, a0, a1, a2;
+    hal_float_t sampleRate, w0, alpha;
```
in line 132 "period" is implicitly converted to hal_float_t. A typecast should be added there.

strahlex commented on src/hal/i_components/blend.icomp:
```diff
@@ -30,7 +30,7 @@ license "GPL";
+    hal_float_t select_ = select;
     if(!open)
         {
         if(select_ < 0) select_ = 0;
```
float is compared 0 should be 0.0, 1 should be 1.0

strahlex commented on src/hal/i_components/ddt.icomp:
```diff
@@ -19,12 +19,12 @@ component ddt "Compute the derivative of the input function";
 ;;
-double tmp = in;
+hal_float_t tmp = in;
 out = (tmp - old) / (period * 1e-9);
```
period is implicitly converted from long to floating point variable, typecast to hal_float_t should be added

strahlex commented on src/hal/i_components/edge.icomp:
```diff
@@ -34,16 +34,16 @@ license "GPL";
-        int desired_edge =
+        hal_s32_t rise = new_in && !last_in;
+        hal_s32_t fall = !new_in && last_in;
+        hal_s32_t desired_edge =
```
these 3 should probably be hal_bit_t boolean variables

ArcEye replied:
And the one above too, done

strahlex commented on src/hal/i_components/flipflop.icomp:
```diff
@@ -4,7 +4,7 @@ pin in bit clk "clock, rising edge writes data to out";
 pin in bit reset "when true, force out false; overrides set";
 pin io bit out "output";
-variable int oldclk;
+variable hal_s32_t oldclk;
```
should be hal_bit_t

strahlex commented on src/hal/i_components/flipflop.icomp:
```diff
@@ -12,8 +12,7 @@ license "GPL";
 {
-int c;
-
+hal_s32_t c;
```
should be hal_bit_t

ArcEye commented:
Good spots, all done and runtest amended, I will refresh the PR shortly

strahlex commented on src/hal/i_components/gantry.icomp:
```diff
@@ -51,10 +51,10 @@ pin in  float search-vel "HOME_SEARCH_VEL from ini file";
+variable hal_float_t offset[7] = 0.0;
+variable hal_float_t prev_cmd = 0.0;
+variable hal_s32_t   fb_joint = 0;
+variable hal_s32_t   latching = 0;
```
should be hal_bit_t

ArcEye replied:
Only latching, fb_joint counts up to the number of joints

strahlex commented on src/hal/i_components/gearchange.icomp:
```diff
@@ -10,14 +10,14 @@ pin in float speed_in "Speed command input";
+pin io bit reverse = 0 "Set to 1 to reverse the spindle in second gear";
 
 function _;
 license "GPL";
```
in the float comparisons below 0.0 and 1.0 should be used

strahlex commented on src/hal/i_components/ilowpass.icomp:
```diff
@@ -18,16 +18,16 @@ pin in s32 in;
 towards the input value in each period""";
 
-variable double value;
+variable hal_float_t value;
```
the weird conversion in line 38 should be fixed

ArcEye replied:
Indeed!
As both values multiplied are hal_float_t and I assume the only purpose of casting it is to round off to an int, I will just use (hal_s32_t)
I have no idea why __64 was used

strahlex replied:
Looks like some sort of workaround for a corner case problem in the past. The cast to hal_s32_t should be fine.

strahlex replied:
taking a closer look I see that value is uninitialized at startup meaning that this component may behave differently in some situations. value should be initialized with 0.0 or with the input value at startup. I guess the 0.0 behavior is the one that is wanted in this case.

strahlex commented on src/hal/i_components/invert.icomp:
```diff
@@ -30,7 +30,7 @@ license "GPL";
-    double tmp = in;
+    hal_float_t tmp = in;
     if (deadband < 1e-12) deadband = 1e-12;
     if ( tmp > -deadband && tmp <0)
```
1 und 0 should be replaced by float versions 1.0 and 0.0

strahlex commented on src/hal/i_components/joyhandle.icomp:
```diff
@@ -17,11 +17,11 @@
-param rw bit inverse = 0;
+pin io float power = 2.0;
+pin io float deadband = 0.;
+pin io float scale = 1.;
```
in this file are various instances of 1. and 0. usage. The should be corrected with 1.0 and 0.0. Furthermore, there are more numbers where .0 should be added to numbers to indicate floating point operations.

ArcEye replied:
All as per original, but now changed

strahlex commented on src/hal/i_components/knob2float.icomp:
```diff
@@ -34,8 +34,8 @@ license "GPL";
 {
-    long delta_counts;
-    double tmp_out;
+    hal_float_t delta_counts;
```
should be hal_s32_t, line 45 https://github.com/machinekit/machinekit/pull/694/files#diff-0c5675002ab9f467e09dcb64e1772b6aR45 contains an implicit type conversion that should be marked with a typecast

ArcEye replied:
Presume you mean line 43, all the vars on 45 are hal_float_t
The component is not satisfactory in that respect, has always mixed s32 with float

strahlex replied:
No, take a look at the input pin it is s32 probably some encoder input. Therefore, old_count and delta_count should be s32 too. The conversion happens at line 45 where the delta_counts are scaled and added to the float value.

ArcEye replied:
OK, you are saying that the types they should be would need a typecast, 
I was saying the ones they currently are do not.

Changed to s32 and typecast accordingly

strahlex commented on src/hal/i_components/latencybins.icomp:
```diff
@@ -25,6 +25,10 @@ Usage:
 Maintainers note: hardcoded for MAXBINNUMBER==1000
+
+NB.  Since there is no hal_64_t yet, this component has been left 'as is'
+using __64 type etc
```
rtapi_get_clocks returns long long int, _s64 could be used internally by the component, the rest could still be converted to hal data types

ArcEye replied:
I have done that, but left __32 and __64 alone.

strahlex replied:
I think __32 could be exchanged by hal_s32_t. The two __64 variables are used to store the timestamps which is fine. The __32 variables is used for output only.

ArcEye commented:
PR refreshed to this point

ArcEye commented on src/hal/i_components/near.icomp:
```diff
@@ -1,8 +1,8 @@
 component near "Determine whether two values are roughly equal.";
 pin in float in1_;
 pin in float in2_;
-param rw float scale=1;
```
fixed to 1.0 etc

ArcEye commented on src/hal/i_components/pid.icomp:
```diff
@@ -147,8 +147,8 @@ author "John Kasunich";
+    hal_float_t periodfp, periodrecip;
 
     /* precalculate some timing constants */
     periodfp = (double)period * 0.000000001;
```
This and later assignment cast (hal_float_t) in next refresh

ArcEye commented on src/hal/i_components/thc.icomp:
```diff
@@ -111,7 +111,7 @@ FUNCTION(_)
+            hal_float_t z_diff;
             z_diff = z_pos_in - last_z_in;
             if(z_diff > 0 && offset != 0){ // torch is moving up
                 if(offset > 0){ // positive offset
```
All BJT's float = 0; assignments changed to 0.0

ArcEye commented on src/hal/i_components/thcud.icomp:
```diff
@@ -110,7 +110,7 @@ FUNCTION(_)
+            hal_float_t z_diff;
             z_diff = z_pos_in - last_z_in;
             if(z_diff > 0 && offset != 0){ // torch is moving up
                 removing_offset = 1;
```
All BJT's float = 0; assignments changed to 0.0 in next refresh

ArcEye commented on src/hal/i_components/time.icomp:
```diff
@@ -83,7 +83,7 @@ function _;
 {
-    __u32 totalseconds;
+    hal_u32_t totalseconds;
     if(start && !old_start) totalnsec = 0;
```
old_start changed to hal_bit_t, 0 to 0.0 and some pedantic typecasting to (hal_u32_t) in next refresh

ArcEye commented on src/hal/i_components/updown.icomp:
```diff
@@ -3,13 +3,13 @@ pin in bit countup "Increment count when this pin goes from 0 to 1";
+pin io s32 min "If clamp or wrap is set, count will never be less than this number";
+variable hal_s32_t oldup;
+variable hal_s32_t olddown;
+variable hal_s32_t first = 1;
```
hal_s32_t changed to hal_bit_t in next refresh, only ever hold 0 or 1

ArcEye commented:
Hopefully caught remaining similar clean ups in this refresh, but who knows!

strahlex commented on src/hal/i_components/joyhandle.icomp:
```diff
@@ -50,51 +50,51 @@ license "GPL";
-else if ((in <= deadband) && (in >= (-1*deadband))) out = 0.;
-else if (in <= -1.) out = -scale - offset;
-else if ((in > deadband) && (in < 1.))
+else if (deadband <= 00) deadband = 0.0;
```
There is a dot missing in the first number.

strahlex commented on src/hal/i_components/knob2float.icomp:
```diff
@@ -34,13 +34,13 @@ license "GPL";
-    delta_counts = counts - old_counts;
+    delta_counts = ( (hal_float_t)counts - old_counts);
     if ( enable ) {
 	tmp_out = old_out + delta_counts * scale;
```
Should be `tmp_out = old_out + (hal_float_t)delta_counts * scale;`

strahlex commented on src/hal/i_components/limit3.icomp:
```diff
@@ -19,8 +19,8 @@ license "GPL";
+    hal_float_t lin, lout, dt, in_v, min_v, max_v, ramp_a, avg_v, err, dv, dp;
+    hal_float_t min_out, max_out, match_time, est_in, est_out;
 
     /* apply first order limit */
```
line 36: 0 to 0.0
line 41: period should be explicitly cast to hal_float_t

strahlex commented on src/hal/i_components/multiclick.icomp:
```diff
@@ -156,7 +155,7 @@ typedef enum {
 FUNCTION(_)
 {
-    int new_in = in;
+    hal_s32_t new_in = in;
```
should be hal_bit_t

strahlex commented on src/hal/i_components/mux16.icomp:
```diff
@@ -44,20 +44,20 @@ pin in bit sel#[4];
-variable int running;
+variable hal_float_t delaytime;
+variable hal_u32_t lastnum;
+variable hal_u32_t running;
```
should be hal_bit_t

strahlex commented on src/hal/i_components/mux16.icomp:
```diff
@@ -44,20 +44,20 @@ pin in bit sel#[4];
-variable int lastnum;
-variable int running;
+variable hal_float_t delaytime;
+variable hal_u32_t lastnum;
```
should be hal_s32_t

strahlex commented on src/hal/i_components/mux16.icomp:
```diff
@@ -44,20 +44,20 @@ pin in bit sel#[4];
 {
-    int i,num = 0;
-    int internal[4];
+    hal_u32_t i,num = 0;
```
should be hal_s32_t

strahlex commented on src/hal/i_components/mux16.icomp:
```diff
@@ -44,20 +44,20 @@ pin in bit sel#[4];
-    int i,num = 0;
-    int internal[4];
+    hal_u32_t i,num = 0;
+    hal_u32_t internal[4];
```
should be hal_bit_t or hal_s32_t

strahlex commented on src/hal/i_components/orient.icomp:
```diff
@@ -7,7 +7,7 @@ pin in  float angle       "orient target position in degrees, 0 <= angle < 360";
 pin out float poserr      "in degrees - aid for PID tuning";
 
-variable int   last_enable = 0;
+variable hal_s32_t   last_enable = 0;
```
should be hal_bit_t

strahlex commented on src/hal/i_components/select8.icomp:
```diff
@@ -9,7 +9,7 @@ license "GPL";
 {
     hal_s32_t temp_sel;
-    int i, temp_enable;
+    hal_s32_t i, temp_enable;
```
temp_enable should be hal_bit_t

strahlex commented on src/hal/i_components/stats.icomp:
```diff
@@ -5,9 +5,9 @@ pin in bit enable "";
-variable int prev_enable;
-variable float m2;
-variable float delta;
+variable hal_s32_t prev_enable;
```
should be hal_bit_t

strahlex commented on src/hal/i_components/stats.icomp:
```diff
@@ -5,9 +5,9 @@ pin in bit enable "";
+variable hal_float_t delta;
 function _;
 license "GPL";
 ;;
```
implicit typecast of n in line 26 and 28 should be explicitly cast to hal_float_t

strahlex commented on src/hal/i_components/time.icomp:
```diff
@@ -72,8 +72,8 @@ pin out u32 minutes "Minutes";
 // Global Variables
-variable double totalnsec;
-variable int old_start;
+variable hal_float_t totalnsec;
```
since this values is used as internal time reference it should be long

strahlex commented on src/hal/i_components/time.icomp:
```diff
@@ -83,15 +83,15 @@ function _;
 
     if(start){
     totalnsec = totalnsec + period;
     totalseconds = totalnsec * 0.000000001;
```
Implicit typecast to floating point value is happening here Should be replaced by integer division.
`totalseconds = (hal_u32_t)(totalnsec / 1e9)`

strahlex commented on src/hal/i_components/time.icomp:
```diff
@@ -83,15 +83,15 @@ function _;
-    __u32 totalseconds;
-    if(start && !old_start) totalnsec = 0;
+    hal_u32_t totalseconds;
+    if(start && !old_start) totalnsec = 0.0;
```
totalnsec should be long therefore this number should be 0

strahlex commented on src/hal/i_components/toggle.icomp:
```diff
@@ -1,10 +1,10 @@
-variable int debounce_cntr;
-variable int debounced;
+variable hal_u32_t debounce_cntr;
+variable hal_u32_t debounced;
```
should be hal_bit_t

strahlex commented on src/hal/i_components/toggle2nist.icomp:
```diff
@@ -14,8 +14,8 @@ pin in  bit in;
 pin out bit off;
-variable int old_in;
-variable int to_state=0;
+variable hal_s32_t old_in;
```
should be hal_bit_t

strahlex commented on src/hal/i_components/toggle2nist.icomp:
```diff
@@ -14,8 +14,8 @@ pin in  bit in;
-variable int old_in;
-variable int to_state=0;
+variable hal_s32_t old_in;
+variable hal_s32_t to_state=0;
```
should be hal_bit_t

strahlex commented:
Well done. This is a huge cleanup. I added a few more comments.

ArcEye commented:
Updated PR, now hopefully covers all previous
Had to amend the ilowpass runtest, as changing the pin types produces a slightly different output on the second test (out by -000002).
Given what it is generally used for (smoothing MPGs etc) should make no practical difference to anyone.

This has tidied code going back to the Ark.  char or int has been used to hold bit or bool values in every component and some people always write their code with unqualified floating points.

I am thinking of specifically defining bool within instcomp, so that variables can be defined as bool
and values set with true or false.
It will not save any space, but will make it far clearer what is going on in the code.
I believe you had similar thoughts?
@strahlex  bool undefined when using comp in RIP installs #635 

My grep and sed expedition only looked at components where param or variable was declared
There will still be float and int types used within the function bodies of some components

That is for another time:)

strahlex commented:
@ArcEye Yes, using true and false for boolean values is the correct way. I have checked again and I get no error when using true, false or TRUE and FALSE on my local RIP installation. Either the problem has been fixed and `<stdbool.h>` is now included or I discovered the problem on another platform.

strahlex commented on src/hal/i_components/ilowpass.icomp:
```diff
@@ -18,16 +18,16 @@ pin in s32 in;
 filter of \\fBgain\\fR per period.""";
 
-param rw float scale = 1024 """A scale factor applied to the output
+pin io float scale = 1024 """A scale factor applied to the output
```
1024 should be replace with floating point equivalent 1024.0

strahlex commented on src/hal/i_components/ilowpass.icomp:
```diff
@@ -35,6 +35,6 @@ license "GPL";
 author "Jeff Epler <jepler@unpythonic.net>";
 ;;
 value += (in - value) * gain;
```
`in` should be typecast to `hal_float_t`

strahlex commented:
Not sure what the behavior of the ilowpass component means. That the test results differ after the changes is more than weird.

ArcEye commented:
The ilowpass behavior is directly related to the intermediate typecast to __s64

If you reinstate the original logic, but now using hal_s32_t instead of int so that it reads
`out = (hal_s32_t)(__s64)(value * scale);`

The output from runtests is back as it was, ie.
```
Running test: tests/ilowpass.0
--- tests/ilowpass.0/expected	2015-06-24 13:06:06.000000000 +0100
+++ tests/ilowpass.0/result	2015-06-25 09:37:14.871841033 +0100
@@ -1,2 +1,2 @@
 2147400000
--2147483648          (this was the output when __s64 was omitted, runtest changed to reflect)
+-2147467296          (this was original output, which is now restored)
*** tests/ilowpass.0: FAIL: result differed from expected
Runtest: 1 tests run, 0 successful, 1 failed + 0 expected
Failed: 
    tests/ilowpass.0
```
Jeff Epler may be a lot of things, but stupid is not one of them.
There is obviously a need to use all the available digits in __s64 to get the right figure when it is rounded off to __s32

I have reverted it to 'as was' and reverted the runtest expected to previous figure

The other points re correction of values to 0.0 etc have been done

PR will be refreshed accordingly soon.

ArcEye commented:
@strahlex  - Second commit joined to this PR as discussed

strahlex commented on src/hal/i_components/biquad.icomp:
```diff
@@ -122,14 +122,14 @@ Biquad_CalcCoeffs(void *arg, unsigned long period)
         valid = 0;
 
-        sampleRate = 1.0 / (period * 1e-9);
+        sampleRate = (hal_float_t) (1.0 / (period * 1e-9));
```
better `sampleRate = (1.0 / ((hal_float_t)period * 1e-9));` prevents implicit casting of the long value

strahlex commented:
Rest looks good now.

ArcEye commented:
Second commit updated and refreshed, should all be there now
